### PR TITLE
8269481: SctpMultiChannel never releases own file descriptor

### PR DESCRIPTION
**Pull request by bplb**
Please review this change to the Unix implementations of `sun.nio.ch.sctp.Sctp*ChannelImpl#kill()` to close the socket if `state == ChannelState.UNINITIALIZED`.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269481](https://bugs.openjdk.java.net/browse/JDK-8269481): SctpMultiChannel never releases own file descriptor

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to 8ca5b5c086e3c05682f47908e79a126df36dff8d
 * [Chris Hegarty](https://openjdk.java.net/census#chegar) (@ChrisHegarty - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4621/head:pull/4621` \
`$ git checkout pull/4621`

Update a local copy of the PR: \
`$ git checkout pull/4621` \
`$ git pull https://git.openjdk.java.net/jdk pull/4621/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4621`

View PR using the GUI difftool: \
`$ git pr show -t 4621`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4621.diff">https://git.openjdk.java.net/jdk/pull/4621.diff</a>

</details>

### CODE REVIEW

**Comment by bplb** [comment]
It would be good if someone more familiar with the SCTP channels could take a look at this one.

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java
Comment (AlanBateman) [inline src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java:643]: It might be better to invert these, meaning set the state to KILL before close(fdVal), just in case the close throws.

Comment (bplb) [reply src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java:643]: So modified.

Comment (ChrisHegarty) [inline src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java:643]: The change here seems mostly benign - there is no reason why either SctpChannel or SctpServerChannel would find themselves in this state - but it seems ok.

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java
Comment (ChrisHegarty) [inline src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java:377]: Good catch. It seems that there was a missing state here, but not worth adding at this point.

File: test/jdk/com/sun/nio/sctp/SctpMultiChannel/CloseDescriptors.java
Comment (ChrisHegarty) [inline test/jdk/com/sun/nio/sctp/SctpMultiChannel/CloseDescriptors.java:71]: Is this test stable? Do we do similar `losf` in other areas ?

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java
Comment (bplb) [reply src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java:377]: I also thought that it looks like there is a missing state.

File: test/jdk/com/sun/nio/sctp/SctpMultiChannel/CloseDescriptors.java
Comment (bplb) [reply test/jdk/com/sun/nio/sctp/SctpMultiChannel/CloseDescriptors.java:71]: `test/jdk/java/net/DatagramSocket/UnreferencedDatagramSockets.java
`
`test/jdk/java/net/MulticastSocket/UnreferencedMulticastSockets.java
`

Might want to change this test to be robust as to the location of the `lsof` executable.

**Comment by bplb** [comment]
If this looks all right could I get an approval please?

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java
Comment (AlanBateman) [reply src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java:643]: It looks like you've changed some but not all cases. I realize ChannelState.UNINITIALIZED is not too interesting but further maintainers may wonder about it.

**Review by AlanBateman — APPROVED** [review]
The implementation changes in the latest round look fine.

I didn't study the test too closely I wonder about the reliability of using a hard coded port (12345) and using losf on an agent VM. I think it will minimally need to run in /othervm mode to avoid shimmer in the numbest of open file descriptors that might arise due to the tests that ran previously in the same VM.

**Review by AlanBateman — APPROVED** [review]
Thanks for the update. You should probably wait for Chris as he had comments on this too.

**Comment by bplb** [comment]
Sure, no problem.

File: test/jdk/com/sun/nio/sctp/SctpMultiChannel/CloseDescriptors.java
Comment (jaikiran) [inline test/jdk/com/sun/nio/sctp/SctpMultiChannel/CloseDescriptors.java:92]: Hello Brian,

The if condition checks for `> 3` whereas the error message states `> 2`. Should the error message be changed? Also, I see that there's a `MAX_DESC` constant in this test, but from what I can see isn't used anywhere? Perhaps that's the constant you wanted to use in this if block and the error message?

**Review by ChrisHegarty — APPROVED** [review]

**Pull request closed without merging**
